src/main/java/net/minecraftforge/common/plants/IGrowablePlant.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2018.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.plants;
+
+import java.util.Random;
+
+import net.minecraft.block.IGrowable;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.IBlockReader;
+import net.minecraft.world.World;
+
+/**
+ * The base interface of all growable plants. Implementers must override canGrow/grow, or callers will end up with a StackOverflowException. This class extends IGrowable in order
+ * to reduce the number of patches needed. Modders should not override the IGrowable methods.
+ */
+@SuppressWarnings("deprecation")
+public interface IGrowablePlant extends IPlant, IGrowable
+{
+
+    /**
+     * @param world
+     *            The world
+     * @param pos
+     *            The current pos
+     * @param state
+     *            The current state
+     * 
+     * @return If this plant can grow, and {@link IPlant#grow(World, Random, BlockPos, IBlockState)} will actually do something.
+     */
+    default boolean canGrow(World world, BlockPos pos, IBlockState state)
+    {
+        return canGrow(world, pos, state, world.isRemote);
+    }
+
+    /**
+     * @param world
+     *            The world
+     * @param rand
+     *            A random
+     * @param pos
+     *            The current pos
+     * @param state
+     *            The current state
+     * 
+     * @return If this plant can be forcibly grown using bonemeal.
+     */
+    boolean canUseBonemeal(World world, Random rand, BlockPos pos, IBlockState state);
+
+    /**
+     * This method is called to grow the plant. If this plant cannot be grown further, this method should do nothing. This method is not responsible for posting the forge
+     * CropGrowPre/Post events. Callers should post those themselves.
+     * 
+     * @param world
+     *            The world
+     * @param pos
+     *            The current pos
+     * @param state
+     *            The current state
+     * @param natural
+     *            If this grow call was a result of "natural" causes like random block updates, or artificial, from fertilizers such as bonemeal.
+     */
+    default void grow(World world, Random rand, BlockPos pos, IBlockState state, boolean natural)
```

Comment:
1) Why is this hard world object, we're trying to move to the interfaces, so use IBlockReader/IBlockWriter. Not the concrete class.
2) What does natural matter here?
3) IBlockState goes as the first argument. And if this is FORCED to be on Blocks, why not just make it sub IForgeBlock?
4) Why are you providing two versions of every function? Just provide the once. Modder can be unlazy.

Reply:
1) Moved to IWorld
2) Some plants may want to differ based on natural growth and forced growth. I.E. a plant that can grow into two different variants if forced or natural.
3) World was kept as first argument to be consistent with vanilla, but that can be changed if necessary.
4) That was due to IGrowable inheritance which has since been changed.
5) Can't since vanilla doesn't.  Checking canGrow beforehand should be acceptable to know if the events should be called.  Call chains should basically look like `if(canGrow){ preGrow() grow() postGrow()}`